arraycontext/impl/pytato/utils.py

```diff
@@ -124,4 +126,60 @@ def get_loopy_target(self) -> "lp.PyOpenCLTarget":
 
 # }}}
 
+# {{{ Transfer mappers
+
+
+class TransferToDeviceMapper(CopyMapper):
+    def __init__(self, queue: CommandQueue, allocator: AllocatorBase = None) -> None:
+        super().__init__()
+        self.queue = queue
+        self.allocator = allocator
+
+    def map_data_wrapper(self, expr: DataWrapper) -> Array:
+        import numpy as np
+
+        import arraycontext.impl.pyopencl.taggable_cl_array as tga
+        if isinstance(expr.data, np.ndarray):
+            data = tga.to_device(self.queue, expr.data, allocator=self.allocator)
+            return DataWrapper(
+                data=data,
+                shape=expr.shape,
+                axes=expr.axes,
+                tags=expr.tags,
+                non_equality_tags=expr.non_equality_tags)
+
+        return super().map_data_wrapper(expr)
+
+
+class TransferToHostMapper(CopyMapper):
+    def __init__(self, queue: CommandQueue, allocator: AllocatorBase = None) -> None:
+        super().__init__()
+        self.queue = queue
+        self.allocator = allocator
+
+    def map_data_wrapper(self, expr: DataWrapper) -> Array:
+        import arraycontext.impl.pyopencl.taggable_cl_array as tga
+        if isinstance(expr.data, tga.TaggableCLArray):
+            data = expr.data.get()
+            return DataWrapper(
```

Comment:
@majosm asked in https://github.com/inducer/pytato/pull/548#discussion_r1773622516

> Can we make this return an instance of a new DataWrapper type that can __hash__ by data instead of id? That would remove the need for data wrapper deduplication, if I understand the intended use of these mappers correctly.

Reply:
Not sure; in that case, we probably should probably also override the `EqualityComparer`, otherwise we would get a lot of hash collisions when caching.